From 97e305025e987e77da488133ded31c1e81a0282b Mon Sep 17 00:00:00 2001
From: Dominique Martinet <dominique.martinet@atmark-techno.com>
Date: Wed, 16 Nov 2022 07:08:13 +0900
Subject: [PATCH] sed: check errors writing file with sed -i
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

sed would currently not error if write failed when modifying a file.

This can be reproduced with the following 'script':
$ sudo mount -t tmpfs tmpfs -o size=1M /tmp/m
$ sudo chmod 777 /tmp/m
$ echo foo > /tmp/m/foo
$ dd if=/dev/zero of=/tmp/m/fill bs=4k
dd: error writing '/tmp/m/fill': No space left on device
256+0 records in
255+0 records out
1044480 bytes (1.0 MB, 1020 KiB) copied, 0.00234567 s, 445 MB/s
$ busybox sed -i -e 's/.*/bar/' /tmp/m/foo
$ echo $?
0
$ cat /tmp/m/foo
<empty>

new behaviour:
$ echo foo > /tmp/m/foo
$ ./busybox sed -i -e 's/.*/bar/' /tmp/m/foo
sed: write error
$ echo $?
4
$ cat /tmp/m/foo
foo

function                                             old     new   delta
sed_main                                             754     801     +47
------------------------------------------------------------------------------
(add/remove: 0/0 grow/shrink: 1/0 up/down: 47/0)               Total: 47 bytes
   text	   data	    bss	    dec	    hex	filename
  66957	   2398	   1552	  70907	  114fb	busybox_old
  67004	   2398	   1552	  70954	  1152a	busybox_unstripped

Signed-off-by: Dominique Martinet <dominique.martinet@atmark-techno.com>
---
 editors/sed.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/editors/sed.c b/editors/sed.c
index 00dde60be..6179c5e80 100644
--- a/editors/sed.c
+++ b/editors/sed.c
@@ -1648,6 +1648,11 @@ int sed_main(int argc UNUSED_PARAM, char **argv)
 			fchown(nonstdoutfd, statbuf.st_uid, statbuf.st_gid);
 
 			process_files();
+			fflush(G.nonstdout);
+			if (ferror(G.nonstdout)) {
+				xfunc_error_retval = 4;  /* It's what gnu sed exits with... */
+				bb_simple_error_msg_and_die(bb_msg_write_error);
+			}
 			fclose(G.nonstdout);
 			G.nonstdout = stdout;
 
